docs: Adds multiturn simulation howto guide (#786)
langchain-ai/langsmith-docs: docs/evaluation/how_to_guides/multi_turn_simulation.mdx

import {
  CodeTabs,
  python,
  typescript,
} from "@site/src/components/InstructionsWithCode";

# How to simulate multi-turn interactions

:::info Key concepts
- [Multi-turn interactions](../concepts#multi-turn-interactions)
- [Evaluators](../concepts#evaluators)
- [LLM-as-judge](../concepts#llm-as-judge)
- [OpenEvals](https://github.com/langchain-ai/openevals)
:::

Some AI applications operate over multiple interactions with a user, also called conversation *turns*.

> When evaluating the performance of such applications, core concepts such as [gathering a dataset](../../evaluation/concepts#datasets) and defining

Author: I'm fine with this but feels pedantic?

> [evaluators](../../evaluation/concepts#evaluators) to judge results successful remain useful - the inputs and outputs for these

Author: Ah yeah that's right

```bash
export LANGSMITH_TRACING="true"
export LANGSMITH_API_KEY="<Your LangSmith API key>"
export OPENAI_API_KEY="<Your OpenAI API key>"
```

## Running a simulation

There are two primary components you'll need to get started:

- `app`: Your application, or a function wrapping it. Must accept a single chat message (dict with "role" and "content" keys) as an input arg and a `thread_id` as a kwarg.
Should accept other kwargs as more may be added in future releases. Returns a chat message as output with at least role and content keys.
- `user`: The simulated user. In this guide, we will use an imported prebuilt function named `create_llm_simulated_user` which uses an LLM to generate user responses, though you can [create your own too](https://github.com/langchain-ai/openevals?tab=readme-ov-file#custom-simulated-users).

The simulator in `openevals` passes a single chat message to your `app` from the `user` for each turn.
Therefore you should statefully track the current history internally based on `thread_id` if needed.

Here's an example that simulates a multi-turn interaction with a simple chat app that wraps a single call to the OpenAI chat completions API.

> Our app is a stand-in for a customer support agent, and our simulated user is playing the role of a particularly aggressive customer:

Reviewer: ```suggestion
In this example, our simulated user is playing the role of a particularly aggressive customer:
```
